File: Python/Repeated DNA Sequences.py
class Solution:
    # @param s, a string
    # @return a list of strings
    def findRepeatedDnaSequences(self, s):
        d = {}
        for i in range(len(s) - 9):
            st = s[i:i+10]
            if st in d:
                d[st] += 1
            else:
                d[st] = 1
        res = []
        for item in d:
            if d[item] >= 2:
                res.append(item)
        return res

File: C++/Repeated-DNA-Sequences.cpp
class Solution {
public:
	int convertInt(const string& str, map<char, int>& hash) {
		int res = 0;
		for (int i = 0; i < 10; i ++) {
			res = res * 4 + hash[str[i]];
		}
		return res;
	}
    vector<string> findRepeatedDnaSequences(string s) {
        vector<string> res;
        map<char, int> hash;
        map<int, int> vis;
        int size = 1 << 20;
        hash['A'] = 0, hash['C'] = 1, hash['G'] = 2, hash['T'] = 3;
        for (int i = 0; i + 9 < s.size(); i ++) {
        	string str = s.substr(i, 10);
        	int cur = convertInt(str, hash);
        	vis[cur] ++;
        	if (vis[cur] == 2) {
        		res.push_back(str);
        	}
        }
        return res;
    }
};